### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by andrii-koptiev**
[DEMO LINK](https://andrii-koptiev.github.io/react_dynamic-list-of-goods/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,12 +1,106 @@
...
+    loadingFive: false,
+    loadingRed: false,
+  };
+
+  getAllGoods = () => {
+    this.setState({ loadingAll: true });
+    getAll()
+      .then(goods => {
+        this.setState({ goods });
+        this.setState({ loadingAll: false });
```
Comment (Rozdorozhnii) [inline src/App.tsx:29]: you can do it in one setState statement

File: src/App.tsx [hunk 2]
```diff
@@ -1,12 +1,106 @@
...
+      });
+  };
+
+  getFiveFirst = () => {
+    this.setState({ loadingFive: true });
+    get5First()
+      .then(goods => {
+        this.setState({ goods });
+        this.setState({ loadingFive: false });
```
Comment (Rozdorozhnii) [inline src/App.tsx:38]: same

File: src/App.tsx [hunk 3]
```diff
@@ -1,12 +1,106 @@
 import React from 'react';
+import classNames from 'classnames';
+import 'bulma';
```
Comment (Rozdorozhnii) [inline src/App.tsx:3]: better to import components which you'll use

**Review by Rozdorozhnii — APPROVED** [review]
Cool using Bulma👍

File: src/App.tsx [hunk 4]
```diff
@@ -1,12 +1,103 @@
...
+      });
+  };
+
+  getRedGoods = () => {
+    this.setState({ loadingRed: true });
+    getRedGoods()
+      .then(goods => {
+        this.setState({ goods, loadingRed: false });
+      });
+  };
```
Comment (Elena-brv) [inline src/App.tsx:46]: try to create a common method for any button

File: src/App.tsx [hunk 5]
```diff
@@ -1,12 +1,103 @@
...
+          </div>
+          {goods.length > 0
+            && !loadingAll
+            && !loadingFive
+            && !loadingRed
+            && <GoodsList goods={goods} />}
+        </section>
+      </div>
+
+    );
```
Comment (Elena-brv) [inline src/App.tsx:99]:
```suggestion
      </div>
    );
```

File: src/GoodsList.tsx [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+      {goods.map(good => (
+        <tr key={good.id}>
+          <th>{good.id}</th>
+          <td style={{ color: good.color }}>{good.name}</td>
+          <td>{good.color}</td>
+        </tr>
+      ))}
+    </tbody>
+
+  </table>
```
Comment (Elena-brv) [inline src/GoodsList.tsx:26]:
```suggestion
    </tbody>
  </table>
```

File: src/api/goods.ts [hunk 7]
```diff
@@ -6,6 +6,16 @@ export function getAll(): Promise<Good[]> {
     .then(response => response.json());
 }
 
-export const get5First = () => {};
+export const get5First = () => {
+  return getAll()
+    .then(goods => {
+      return [...goods].sort((a, b) => a.name.localeCompare(b.name)).slice(0, 5);
+    });
```
Comment (Elena-brv) [inline src/api/goods.ts:13]:
```suggestion
    .then(goods => goods
      .sort((a, b) => a.name.localeCompare(b.name))
      .slice(0, 5);
    );
```

File: src/api/goods.ts [hunk 8]
```diff
@@ -6,6 +6,16 @@ export function getAll(): Promise<Good[]> {
...
+};
 
-export const getRedGoods = () => {};
+export const getRedGoods = () => {
+  return getAll()
+    .then(goods => {
+      return goods.filter(good => good.color === 'red');
+    });
```
Comment (Elena-brv) [inline src/api/goods.ts:20]:
```suggestion
    .then(goods => goods.filter(good => good.color === 'red'));
```

**Review by Elena-brv — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 9]
```diff
@@ -1,12 +1,102 @@
...
+class App extends React.Component<{}, State> {
+  state: State = {
+    goods: [],
+    loadingAll: false,
+    loadingFive: false,
+    loadingRed: false,
+  };
+
+
+  loadGoods = async (event: React.MouseEvent<HTMLButtonElement>, callback: Callback) => {
```
Comment (Elena-brv) [inline src/App.tsx:25]:
```suggestion
  };

  loadGoods = async (event: React.MouseEvent<HTMLButtonElement>, callback: Callback) => {
```

File: src/App.tsx [hunk 10]
```diff
@@ -1,12 +1,102 @@
...
+
+    this.setState({ [name]: true } as unknown as { [K in keyof State]: State[K] });
+
+    const goods = await callback();
+
+    this.setState({
+      goods: [...goods],
+      [name]: false
+    } as unknown as { [K in keyof State]: State[K]
+      })
```
Comment (Elena-brv) [inline src/App.tsx:36]:
```suggestion
    this.setState({
      goods: [...goods],
      [name]: false
    } as unknown as { [K in keyof State]: State[K] })
```

**Review by Elena-brv — APPROVED** [review]

**Review by Elena-brv — APPROVED** [review]

**Pull request closed without merging**